Add support for deletion of group offsets to rpk

graphcareful opened the pull request:
This change adds a new command to rpk, under the `group` meta-command called `offset-delete` which will allow users to delete group offsets. Topic/Partitions can be passed to the command in two ways. 

1. Via the `--from-file` option. Taking an input a filename which is expected to be formatted to contain a list of topic partition tuples, and a group name via command line args.
```
$ cat input.txt
foo 0
bar 0
bar 1
bar 2

$ rpk group offset-delete my-consumer-group1 input.txt
bar  1     UNKNOWN_TOPIC_OR_PARTITION: This server does not host this topic-partition.
bar  0     UNKNOWN_TOPIC_OR_PARTITION: This server does not host this topic-partition.
bar  2     UNKNOWN_TOPIC_OR_PARTITION: This server does not host this topic-partition.
foo  0     OK
```

2. Via the `-topic (-t)` option. This option can be used repeatedly to pass topics or topic/partitions directly to the command. If only topics are used a metadata request will be made to find all the partitions for the topic, alleviating the need for the user to know the number of partitions for a topic if they just want to remove all offsets for a given consumer groups topic:
```
$ rpk group offset-delete my-consumer-group1 -topic foo -topic bar:0,1,2
bar  1     UNKNOWN_TOPIC_OR_PARTITION: This server does not host this topic-partition.
bar  0     UNKNOWN_TOPIC_OR_PARTITION: This server does not host this topic-partition.
bar  2     UNKNOWN_TOPIC_OR_PARTITION: This server does not host this topic-partition.
foo  0     OK
```

## Backports Required

- [ ] none - not a bug fix
- [x] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v22.3.x
- [ ] v22.2.x
- [ ] v22.1.x

## UX Changes

* New command `rpk group offset-delete` will be added to `rpk`

## Release Notes
  ### Features

* New command `rpk group offset-delete` will be added to `rpk`

twmb commented on src/go/rpk/pkg/cli/cmd/group/offset_delete.go:
```diff
@@ -0,0 +1,114 @@
+	}
+}
+
+func parseTopicPartitionsFile(fs afero.Fs, file string) (kadm.TopicsSet, error) {
```
This can be replaced with `out.ParseFileArray`, i.e.,

type offsetDeleteRow {
  topic string
  partition int32
}

out.ParseFileArray[offsetDeleteRow](fs, file)
```

graphcareful commented:
Converting to draft as this will now depend on changes in https://github.com/redpanda-data/redpanda/pull/8669 which is currently under review

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/cd24c8d922e3ad5cd62f1b45fb1a3162469542be..2695a7098170a5309f82eab1789162d90dcf4dc4)
- Rebased against dev to resolve merge conflict

twmb commented on src/go/rpk/pkg/out/in.go:
```diff
@@ -137,3 +137,31 @@ func parseSpaceLines[T any](r io.Reader) ([]T, error) {
+	tps := make(map[string][]int32)
+	for _, item := range list {
+		split := strings.SplitN(item, ":", 2)
+		if len(split[0]) == 0 {
+			return nil, fmt.Errorf("item %q invalid empty topic", item)
+		}
```
This conditional can be deleted, the conditional right below is doing what we want (and this conditional was meant to be that one).

twmb commented on src/go/rpk/pkg/cli/cmd/group/offset_delete.go:
```diff
@@ -0,0 +1,147 @@
+
+			// Parse input from file or command line
+			var topicsSet kadm.TopicsSet
+			if len(topicPartitions) > 0 {
+				topicsSet, err = parseTopicPartitionsArgs(topicPartitions)
+			} else if fromFile != "" {
```
Above this, can we have a conditional that exits if len(topicPartitions) > 0 && fromFile != ""?

twmb replied:
oh nice, I missed that

twmb commented on src/go/rpk/pkg/cli/cmd/group/offset_delete.go:
```diff
@@ -0,0 +1,147 @@
+			// For -topic options that didn't include partitions, perform a
+			// lookup for them using a metadata request
+			var m kadm.Metadata
+			m, err = adm.Metadata(context.Background())
```
I think this would be better to only issue metadata for topics that are empty in topicsSet, rather than getting metadata for the whole cluster. This would also simplify fillTopicsSet because we can always use all returned partitions for all topics.

graphcareful replied:
Nice advice, this removes a lot of code, becomes real succinct now:
```
			resp, err := adm.Metadata(context.Background(), topicsSet.EmptyTopics()...)
			out.MaybeDie(err, "unable to make metadata request: %v", err)
			topicsSet.Merge(resp.Topics.TopicsSet())
```

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/2695a7098170a5309f82eab1789162d90dcf4dc4..1f535f1db93ef16f876dc8b527696300cd4e3099)
- Edited the metadata request logic to only request metadata for topics that the user listed without partitions
- Removed an unnecessary conditional in `ParseTopicPartitions`

twmb commented on src/go/rpk/pkg/cli/cmd/group/offset_delete.go:
```diff
@@ -0,0 +1,128 @@
+			out.MaybeDie(err, "unable to make metadata request: %v", err)
+			topicsSet.Merge(mdResp.Topics.TopicsSet())
+
+			// Using kafka-client, invoke broker
```
odd comment here and line 83

twmb commented on src/go/rpk/pkg/cli/cmd/group/offset_delete.go:
```diff
@@ -0,0 +1,128 @@
+			}
+		},
+	}
+	cmd.Flags().StringVarP(&fromFile, "from-file", "f", "", "File of Topic/Partition tuples for which to delete offsets for")
```
lowercase `Topic/Partition`

twmb commented on src/go/rpk/pkg/cli/cmd/group/offset_delete.go:
```diff
@@ -0,0 +1,128 @@
+		},
+	}
+	cmd.Flags().StringVarP(&fromFile, "from-file", "f", "", "File of Topic/Partition tuples for which to delete offsets for")
+	cmd.Flags().StringArrayVarP(&topicPartitions, "topic", "t", nil, "topic:partition_id; (repeatable; e.g. -t foo:0,1,2 )")
```
no semicolon after `partition_id`

twmb reviewed (approved):
lgtm, a few non blocking nits.

graphcareful commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/1f535f1db93ef16f876dc8b527696300cd4e3099..f4e783b1e733554848f599d9334482bc2586381d)
- Addressed nits
